packages/gatsby/src/query/index.js

```diff
@@ -161,6 +161,20 @@ const groupQueryIds = queryIds => {
 const processQueries = async (queryJobs, activity) => {
   const queue = queryQueue.createBuildQueue()
   await queryQueue.processBatch(queue, queryJobs, activity)
+  if (process.env.GATSBY_PAGE_BUILD_ON_DATA_CHANGES) {
+    const { pages } = store.getState()
+    const { pageData } = readState()
+
+    if (pageData) {
+      readState().pageData.forEach((value, key) => {
+        if (!pages.has(key)) {
+          boundActionCreators.removePageData({
+            id: key,
+          })
```

Comment:
Removing page data only in build seems very dangerous (state could get very weird when you mix builds in develop locally without clearing the cache) - we should probably do this as soon as some query result is invalidated - we already use `DELETE_COMPONENTS_DEPENDENCIES` action type / internal `deleteComponentsDependencies` action - so I feel like we could use that action to also clear data in `pageData` substate would make sense?

Reply:
@pieh  
Sure, I've added this additional remove of page data in the `DELETE_COMPONENTS_DEPENDENCIES` reducer, is this correct?

https://github.com/gatsbyjs/gatsby/pull/21523/files#diff-1ce4863868339b89c5b5f502dba637b6R50

Reply:
@pieh  could we call the `removePageData` in the hot reload?

https://github.com/gatsbyjs/gatsby/pull/21523/commits/7060d8d11af4cfd62c9e6f59c346e69dafbc46e9

Reply:
I feel that we could, but there might be some weirdness happening - I would need to manually test that to give better answer. Also this call doesn't actually delete the files (.html, page-data), right? File is deleted later on in the build process? So maybe this desync could cause issues. I will need to think about this more

Reply:
@pieh 
Goods point I'm not familiar with the development mode process.  
I've been testing by removing and editing content in develop mode, then running a build in the markdown_id benchmark, it seems to be ok. 

Yep, the action `removePageData` only removes it from the store, so this will be the pages content hash. As long as the cache `redux.state` is not being updated in development mode I think it should be ok, as the store is pre-populated from the cache state. 🤔

Reply:
> As long as the cache redux.state is not being updated in development mode

We currently actually are updating the file in development:
https://github.com/gatsbyjs/gatsby/blob/ebfa8c39054166587c6ca7213c81993563663deb/packages/gatsby/src/db/index.js#L29-L36

And we start the `autoSave` behaviour after we complete bootstrap and initial query running in `gatsby develop`:

https://github.com/gatsbyjs/gatsby/blob/ebfa8c39054166587c6ca7213c81993563663deb/packages/gatsby/src/commands/develop.ts#L415

So we need to keep this behaviour in mind

Reply:
🤔  
Maybe `pageData` shouldn't be removed during development mode. As the `pageData` should represent the last Gatsby build. 

Am I right in my understanding, the `pages` object is not saved in the redux.state, this is created new in the store each build or develop.

Have you had time to test these cases?

Reply:
> Have you had time to test these cases?

I didn't yet. I'm quite busy with other project, so I might need to delay that to weekend or so.

> Am I right in my understanding, the pages object is not saved in the redux.state, this is created new in the store each build or develop.

Alternatively we could just start saving `pages` in `redux.state`. It wasn't saved before mostly because there was no benefit from it as we didn't have to compare previous state to new one on builds because we were building all pages every time anyway - this features changes that, so this need to be re-evaluated.

> Maybe pageData shouldn't be removed during development mode. As the pageData should represent the last Gatsby build.

Maybe for this PR this is a way to go, but ultimately we will want to be able to reuse whatever is possible from previous "sessions" (I call them sessions but I just mean any work that already had been done on previous `gatsby build` or `gatsby develop`).

We would need to make sure that first build after `gatsby develop` will rebuild everything pretty much if we remove `pageData` (which is kind of what already happens now so that wouldn't be regression). Then the "happy path" for this feature would be limited to only doing consecutive `gatsby build`s (which is fine, but might need clarification at some point)

I didn't spend too much time on the changes yet, I just wanted to try to flag anything that I could quickly identify as potential problem. To do deep dive I would probably need to block few days to fully understand impact of the changes.

Reply:
@pieh thanks really appreciate the insight 🙇. 

> so I might need to delay that to weekend or so.

Working weekends 😱,if any issues arise happily jump on a pairing session to get a better understanding.

Reply:
We did group review on this, so this comment thread no longer apply - we decided that safest options for now (and fastest way to get this in) is to disallow the flag completely in `gatsby develop` and also make sure that that if users start using the flag they should continue to do so - in similar way if user didn't use the flag before and start using we should blow away the cache completely. This should work ok for CI environments where it would continuously use this env var and prevent any local development problems